Enabling MIRI Photometry

taylorbell57 opened the pull request:
This PR enables the use of Eureka! for MIRI photometry.

In particular, this PR:
- Edits miri.py's read function to permit photometry files
- Adds the required flag_bg_phot function to miri.py (just calls NIRCam's function)
- Edits miri.py's read function to better handle simulated data without INTSTART and INTEND header values in the FITS files
- Fixes some minor bugs with plots 3503 (not writing over previous plot and consuming excessive RAM; also wasn't being limited by meta.nplots), 3306 & 3504 (hardcoded vmin and vmax weren't good for all datasets), 3505 (run for all integrations and not limited by meta.nplots)
- Changes the behaviour of meta.skyin and meta.skyout if they are lists. Previously, there were no protections against skyin being larger than skyout, so the new method assumes that there is one skyin value for each skyout value. In otherwords, if skyin=[10,11] and skyout=[15,16], previously the code would've tried 10-15, 10-16, 11-15, and 11-16 annuli while now it will just do 10-15 and 10-16. That's important in the case of skyin=[10,15,20], skyout=[15,20,25], where the old algorithm could do some pretty dangerous things.
- Makes running the b2f.convert_to_e function optional (but run by default if the new meta.convert_to_e parameter is missing). This is important to calculate flux-calibrated stellar photometry.
- Allows a manual first-guess of the centroid position rather than the currently implemented full-frame centroid fit. For the NIRCam photometry, the full-frame fit isn't too far off since the PSF is huge and the star is bright, but for MIRI the star is very small and not much brighter than the background (in some filters at least) and there's also significant non-uniformities in the huge full-frame images which make the full-frame centroiding even worse. Users can now pass in meta.ctr_guess with the manually generated x,y guess for the centroid (using the pixel indices before xwindow and ywindow are applied).
- Adds a new ECF template for MIRI photometry
- Documents all added/changed ECF parameters

This PR also:
- Resolves #490 (Using skyin and skywidth is probably the safer choice though, and if you agree I can switch to the approach I suggested in that issue)
- Resolves #313 (Extends photometry support to MIRI, the second and only other JWST instrument that allows photometry to my knowledge)

I think @sebastian-zieba also had a MIRI photometry branch under development. Depending on how close that is to being ready for a PR, we can either:
1. Merge this PR first and then merge in Sebastian's code when it's ready
2. Merge Sebastian's PR into this branch and then merge this branch into main

kevin218 commented on src/eureka/S3_data_reduction/s3_reduce.py:
```diff
@@ -402,29 +406,38 @@ def reduce(eventlabel, ecf_path=None, s2_meta=None, input_meta=None):
+                        else:
+                            guess = np.array(meta.ctr_guess)[::-1]
+                            trim = np.array([meta.ywindow[0], meta.xwindow[0]])
+                            position = guess - trim
```
I'm not keen on requiring a fixed `position` when a guess is provided.  If the user supplies a guess centroid, they should also have the option to feed the guess into `centerdriver`.

taylorbell57 replied:
Well the guess just replaces the first round of centroiding which was done on the entire frame. For MIRI, centroiding on the entire frame is a really bad idea because the frame is so large and there can easily be multiple sources, and even if there is only one source in the field the centroid gets strongly affected by the very non-uniform background.

To be clear, a second round of centroiding is still done in lines 451-464 in the exact same was as was done before (using a smaller subarray around the guess), just using the manually guessed value instead of the full-frame automatic centroiding. And if no guess is provided, then the code runs the exact same as before.

kevin218 replied:
Ah, ok.  Thanks for clarifying.  Now it makes sense.

taylorbell57 commented on docs/source/ecf.rst:
```diff
@@ -269,6 +269,10 @@ photometry
 ''''''''''
 Only used for photometry analyses. Set to True if the user wants to analyze a photometric dataset.
 
+convert_to_e
```
This isn't a required setting (by default is set to true if not present in the ECF) and is only needed if you want to do flux-calibrated photometry. I can add it to the NIRCam and MIRI photometry ECFs if you want, but I really don't think it should be added to the spectroscopy ECFs since it's likely not overly useful there and more likely to cause confusion than anything else

kevin218 replied:
Please add it to the spectroscopic ECFs.  Today, I've been trying to generate a calibrated stellar spectrum with G395 and has to comment out the conversion line in Stage 3.

taylorbell57 commented on src/eureka/S3_data_reduction/miri.py:
```diff
@@ -56,34 +56,72 @@ def read(filename, data, meta, log):
+        # be changed
+        data.attrs['shdr']['DISPAXIS'] = 1
+
+        # FINDME: make this better for all filters
```
I agree this is a bit clunky, but I think it's good enough for now. One complication to keep in mind though is that the exact wavelength is dependent on the stellar spectrum convolved with the filter's throughput.

taylorbell57 commented on demos/JWST/S3_miri_photometry_template.ecf:
```diff
@@ -0,0 +1,44 @@
+xwindow         [550, 800]  # Horizontal axis as seen in DS9
+dqmask          True        # Mask pixels with an odd entry in the DQ array
+
+# Background parameters
```
The nircam photometry file will also need to be updated

taylorbell57 replied:
So the photometry code's `flag_bg_phot` function currently does sigma clipping along the time axis for each individual background pixel independently. As a result, your full-frame code isn't exactly interchangeable with the photometry code. The `flag_bg` boolean is also a on/off switch, and if set False no background sigma clipping is done. It's not completely clear to me how to integrate your new code if that's what you're wanting to do, but as it stands your PR #502 did not make any changes to the photometry code that mean I need to change this new ECF

kevin218 replied:
I can look into implementing full-frame sigma clipping for photometry once both PRs are merged.

taylorbell57 commented:
Alight, I think I've addressed all your comments (aside from the edits that need to wait for #502 to be merged).

I do, however, want your opinion on whether I should continue using `skyin` and `skyout` as is currently done or instead use `skyin` and `skywidth` to make it impossible to have `skyout < skyin` and to make it a bit easier to loop over different aperture options (using an n*m grid of `skyin` and `skywidth` values instead of having to specify one `skyout` for each `skyin`)

kevin218 reviewed (commented):
Changes look good.  I think using `skywidth` makes the most sense.
